pkg/guacrest/server/server.go

```diff
@@ -99,3 +104,106 @@ func (s *DefaultServer) AnalyzeDependencies(ctx context.Context, request gen.Ana
 		return nil, fmt.Errorf("%v sort is unsupported", request.Params.Sort)
 	}
 }
+
+func (s *DefaultServer) GetPackageInfo(ctx context.Context, request gen.GetPackageInfoRequestObject) (gen.GetPackageInfoResponseObject, error) {
+	logger := logging.FromContext(ctx)
+
+	decodedPurlOrArtifact, err := url.QueryUnescape(request.PurlOrArtifact)
+	if err != nil {
+		logger.Infof("error decoding purl or artifact: %v", err)
+		return gen.GetPackageInfo400JSONResponse{
+			BadRequestJSONResponse: gen.BadRequestJSONResponse{
+				Message: fmt.Sprintf("Invalid identifier: %v", err),
+			},
+		}, nil
+	}
+
+	// Determine if the identifier is a PURL or an artifact
+	var pkgInput *model.PkgInputSpec
+
+	if strings.HasPrefix(decodedPurlOrArtifact, "pkg:") {
+		pkgInput, err = helpers2.PurlToPkg(decodedPurlOrArtifact)
+		if err != nil {
+			return gen.GetPackageInfo400JSONResponse{
+				BadRequestJSONResponse: gen.BadRequestJSONResponse{
+					Message: fmt.Sprintf("Failed to parse PURL: %v", err),
+				},
+			}, nil
+		}
+	} else {
+		splitString := strings.Split(decodedPurlOrArtifact, ":")
+
+		if len(splitString) != 2 {
+			return gen.GetPackageInfo400JSONResponse{
+				BadRequestJSONResponse: gen.BadRequestJSONResponse{
+					Message: fmt.Sprintf("Invalid identifier, artifact should be comprised of <algorithm>:<digest>: %v", decodedPurlOrArtifact),
+				},
+			}, nil
+		}
+
+		occurrence, err := model.Occurrences(ctx, s.gqlClient, model.IsOccurrenceSpec{
+			Artifact: &model.ArtifactSpec{
+				Algorithm: &splitString[0],
+				Digest:    &splitString[1],
+			},
+		})
+		if err != nil {
+			return gen.GetPackageInfo400JSONResponse{
+				BadRequestJSONResponse: gen.BadRequestJSONResponse{
+					Message: fmt.Sprintf("Invalid identifier, artifact should be comprised of <algorithm>:<digest>: %v", decodedPurlOrArtifact),
+				},
+			}, nil
+		}
+
+		convertedOccurrence, ok := occurrence.IsOccurrence[0].Subject.(*model.AllIsOccurrencesTreeSubjectPackage)
+		if !ok {
+			return gen.GetPackageInfo500JSONResponse{
+				InternalServerErrorJSONResponse: gen.InternalServerErrorJSONResponse{
+					Message: fmt.Sprintf("Error converting issoccurrence to model.IsOccurrencesTreeSubjectPackage: %v", decodedPurlOrArtifact),
+				},
+			}, nil
+		}
+
+		pkg := convertedOccurrence.AllPkgTree
+
+		pkgInput = &model.PkgInputSpec{
+			Type:      pkg.Type,
+			Namespace: &pkg.Namespaces[0].Namespace,
+			Name:      pkg.Namespaces[0].Names[0].Name,
+			Version:   &pkg.Namespaces[0].Names[0].Versions[0].Version,
+			Subpath:   &pkg.Namespaces[0].Names[0].Versions[0].Subpath,
+		}
```

Comment:
IIUC, this endpoint exclusively searches for metadata associated with a package -- such that calling the endpoint with an artifact doesn't actually search for metadata by that artifact, and instead searches for metadata by the package associated with that artifact. 

First, metadata directly attached to the artifact (as opposed to the package, #1736 is relevant) isn't surfaced. This is likely a bug, but possibly a UX issue, depending on the intended behavior. 

Secondly, once the above bug is fixed, there is a UX issue - it isn't clear from the API specification that searching by digest `A` may return metadata associated with digest `B` (if artifacts `A` and `B` are both occurrences of the same package).

I suggest defining the intended behavior (w.r.t. this issue) of the endpoint (and maybe adding your ideas here) before implementing. #1867 may be helpful, where the same considerations had to be made.

Reply:
Yeah, this was my mistake. I have now changed the code so that when searching via an artifact, it doesn't just search via the metadata associated with a package.